### PR TITLE
Visualisation of local changes

### PR DESCRIPTION
**Pull request by alexbruy**
Add an action "Show local changes" to the layer context menu toolbar button to open a "Diff viewer" dialog where one can examine local changes made in the project
зображення
If there are no changes in the project, a message bar will be shown, telling that there are no changes which can be visualised.
The "Changes Viewerer" dialog provides a tab-based UI, where each tab represents a local layer with changes. Changes in the layer will be shown on canvas and in the table form (splitted with a splitter).
зображення
Toolbar buttons allow user:
- toggle display of project layers (on by default)
- zoom to full extent of the changes layer
- zoom to selected features of the changes layer
- add currently displayed or all changes layers to project

All changes layers have styling applied and conditional formatting of the attribute table, to make it easier to see which values/features were updated/deleted/inserted.
зображення
зображення

A new button "View Changes" were added to the Sync Project dialog. Clicking on that button will close Sync Project dialog and open Changes Viewer.
зображення

Also add two new Processing tools: one for generating layer containing local changes and another one for generating layer containing changes between two versions of the layer.

### CODE REVIEW

**Comment by alexbruy** [comment]
Requires https://github.com/lutraconsulting/mergin-py-client/pull/134

**Comment by alexbruy** [comment]
@wonder-sk I'm not happy with the current workaround for removing memory layers and updating project. But unfortunately, I can't find other way of doing it.

File: Mergin/processing/algs/extract_local_changes.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+            raise QgsProcessingException("Selected layers has unsupported format. Only GPKG layers are supported.")
+
+        mp = MerginProject(project_dir)
+
+        layer_path = layer.source().split("|")[0]
+        diff_path = get_local_changes(layer_path, mp)
+        feedback.setProgress(5)
+
+        if diff_path is None:
+            raise QgsProcessingException("Failed to retrieve changes, as there is no base file for input layer.")
```
Comment (varmar05) [inline Mergin/processing/algs/extract_local_changes.py:95]: wondering if user would know what that means

File: Mergin/project_status_dialog.py [hunk 2]
```diff
@@ -188,3 +198,24 @@ def show_validation_results(self):
                 html.append(f"<ul>{''.join(items)}</ul>")
 
         self.txtWarnings.setHtml(''.join(html))
+
+    def show_changes(self):
+        if self.mp is None:
+            self.mp = MerginProject(QgsProject.instance().homePath())
```
[the remark this answers is not in the document]
Comment (alexbruy) [reply Mergin/project_status_dialog.py:204]: No, we don't need it. Sync button will be disabled if current project is not a Mergin project.

File: Mergin/processing/algs/extract_local_changes.py (hunk 1 above)
Comment (alexbruy) [reply Mergin/processing/algs/extract_local_changes.py]: Made it shorter and without additional technical details.

File: Mergin/processing/algs/create_diff.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        layer = self.parameterAsVectorLayer(parameters, self.LAYER, context)
+
+        if not check_mergin_subdirs(project_dir):
+            raise QgsProcessingException("Selected directory does not contain a valid Mergin project.")
+
+        if not os.path.normpath(layer.source()).lower().startswith(os.path.normpath(project_dir)):
+            raise QgsProcessingException("Selected layer does not belong to the selected Mergin project.")
+
+        if layer.dataProvider().storageType() != "GPKG":
+            raise QgsProcessingException("Selected layers has unsupported format. Only GPKG layers are supported.")
```
[the remark this answers is not in the document]
Comment (alexbruy) [reply Mergin/processing/algs/create_diff.py]: Done.

**Comment by PeterPetrik** [comment]
Would it be possible to have also processing alg too take 2 geopackage files and show the differences (think of resolving conflict files)?

**Comment by alexbruy** [comment]
> Would it be possible to have also processing alg too take 2 geopackage files and show the differences (think of resolving conflict files)?

Yes, we can implement something like this. Should this tool create diffs for all layers in the GeoPackage or user should select specific layer to process?

**Comment by alexbruy** [comment]
@wonder-sk all comments should be addressed now. May I ask you to take a look when you have time?

**Review by wonder-sk — APPROVED** [review]
Great stuff :tada: ... I am happy with it

**Pull request merged**